File: nodecore-spv/src/test/java/veriblock/admin/temp/grpc/model/LedgerContextTest.java
package veriblock.admin.temp.grpc.model;

import org.junit.Assert;
import org.junit.Before;
import org.junit.Test;
import veriblock.model.LedgerContext;
import veriblock.model.LedgerProofStatus;
import veriblock.model.LedgerValue;

public class LedgerContextTest {

    private LedgerContext addressDoesntExist;
    private LedgerContext addressIsInvalid;
    private LedgerContext addressIsExist;

    @Before
    public void setUp() {
        this.addressDoesntExist = new LedgerContext();
        addressDoesntExist.setLedgerProofStatus(LedgerProofStatus.ADDRESS_DOES_NOT_EXIST);

        this.addressIsInvalid= new LedgerContext();
        addressIsInvalid.setLedgerProofStatus(LedgerProofStatus.ADDRESS_IS_INVALID);

        this.addressIsExist= new LedgerContext();
        addressIsExist.setLedgerProofStatus(LedgerProofStatus.ADDRESS_EXISTS);
    }

    @Test
    public void compareToWhenStartValueIsNull() {
        LedgerContext ledgerContext = new LedgerContext();

        Assert.assertEquals(1, ledgerContext.compareTo(addressDoesntExist));
        Assert.assertEquals(1, ledgerContext.compareTo(addressIsInvalid));
        Assert.assertEquals(1, ledgerContext.compareTo(addressIsExist));
    }

    @Test
    public void compareToWhenStartValueAddressNotExist() {
        LedgerContext ledgerContext = new LedgerContext();
        ledgerContext.setLedgerProofStatus(LedgerProofStatus.ADDRESS_DOES_NOT_EXIST);

        Assert.assertEquals(-1, ledgerContext.compareTo(addressDoesntExist));
        Assert.assertEquals(-1, ledgerContext.compareTo(addressIsInvalid));
        Assert.assertEquals(1, ledgerContext.compareTo(addressIsExist));
    }


    @Test
    public void compareToWhenStartValueAddressIsExist() {
        LedgerContext ledgerContext = new LedgerContext();
        ledgerContext.setLedgerProofStatus(LedgerProofStatus.ADDRESS_EXISTS);

        Assert.assertEquals(-1, ledgerContext.compareTo(addressDoesntExist));
        Assert.assertEquals(-1, ledgerContext.compareTo(addressIsInvalid));
    }


    @Test
    public void compareToWhenStartValueAddressIsExistAndSignatureIndexLess() {
        LedgerContext ledgerContext = new LedgerContext();
        ledgerContext.setLedgerProofStatus(LedgerProofStatus.ADDRESS_EXISTS);
        ledgerContext.setLedgerValue(new LedgerValue(0l, 0l, 1l));

        addressIsExist.setLedgerValue(new LedgerValue(0l, 0l, 0l));

        Assert.assertEquals(-1, ledgerContext.compareTo(addressIsExist));
    }

    @Test
    public void compareToWhenStartValueAddressIsExistAndSignatureIndexMore() {
        LedgerContext ledgerContext = new LedgerContext();
        ledgerContext.setLedgerProofStatus(LedgerProofStatus.ADDRESS_EXISTS);
        ledgerContext.setLedgerValue(new LedgerValue(0l, 0l, 1l));

        addressIsExist.setLedgerValue(new LedgerValue(0l, 0l, 3l));

        Assert.assertEquals(1, ledgerContext.compareTo(addressIsExist));
    }

    @Test
    public void compareToWhenAddressIsExistAndSignatureIndexEqualButUnitsLess() {
        LedgerContext ledgerContext = new LedgerContext();
        ledgerContext.setLedgerProofStatus(LedgerProofStatus.ADDRESS_EXISTS);
        ledgerContext.setLedgerValue(new LedgerValue(100l, 0l, 3l));

        addressIsExist.setLedgerValue(new LedgerValue(0l, 0l, 3l));

        Assert.assertEquals(-1, ledgerContext.compareTo(addressIsExist));
    }

    @Test
    public void compareToWhenAddressIsExistAndSignatureIndexEqualButUnitsMore() {
        LedgerContext ledgerContext = new LedgerContext();
        ledgerContext.setLedgerProofStatus(LedgerProofStatus.ADDRESS_EXISTS);
        ledgerContext.setLedgerValue(new LedgerValue(100l, 0l, 3l));

        addressIsExist.setLedgerValue(new LedgerValue(101l, 0l, 3l));

        Assert.assertEquals(1, ledgerContext.compareTo(addressIsExist));
    }

    @Test
    public void compareToWhenAddressIsExistAndSignatureIndexEqual() {
        LedgerContext ledgerContext = new LedgerContext();
        ledgerContext.setLedgerProofStatus(LedgerProofStatus.ADDRESS_EXISTS);
        ledgerContext.setLedgerValue(new LedgerValue(100l, 0l, 3l));

        addressIsExist.setLedgerValue(new LedgerValue(100l, 0l, 3l));

        Assert.assertEquals(-1, ledgerContext.compareTo(addressIsExist));
    }
}

File: nodecore-spv/src/test/java/veriblock/admin/temp/grpc/model/LedgerContextTest.kt
package veriblock.admin.temp.grpc.model

import io.mockk.mockk
import org.junit.Assert
import org.junit.Before
import org.junit.Test
import org.veriblock.sdk.models.Address
import veriblock.model.LedgerContext
import veriblock.model.LedgerProofStatus
import veriblock.model.LedgerValue

class LedgerContextTest {

    private val dummyAddress: Address = mockk()
    private lateinit var addressDoesntExist: LedgerContext
    private lateinit var addressIsInvalid: LedgerContext
    private lateinit var addressIsExist: LedgerContext

    @Before
    fun setUp() {
        addressDoesntExist = LedgerContext(
            address = dummyAddress,
            ledgerProofStatus = LedgerProofStatus.ADDRESS_DOES_NOT_EXIST
        )
        addressIsInvalid = LedgerContext(
            address = dummyAddress,
            ledgerProofStatus = LedgerProofStatus.ADDRESS_IS_INVALID
        )
        addressIsExist = LedgerContext(
            address = dummyAddress,
            ledgerProofStatus = LedgerProofStatus.ADDRESS_EXISTS
        )
    }

    @Test
    fun compareToWhenStartValueIsNull() {
        val ledgerContext = LedgerContext(
            address = dummyAddress
        )
        Assert.assertEquals(1, ledgerContext.compareTo(addressDoesntExist).toLong())
        Assert.assertEquals(1, ledgerContext.compareTo(addressIsInvalid).toLong())
        Assert.assertEquals(1, ledgerContext.compareTo(addressIsExist).toLong())
    }

    @Test
    fun compareToWhenStartValueAddressNotExist() {
        val ledgerContext = LedgerContext(
            address = dummyAddress,
            ledgerProofStatus = LedgerProofStatus.ADDRESS_DOES_NOT_EXIST
        )
        Assert.assertEquals(-1, ledgerContext.compareTo(addressDoesntExist).toLong())
        Assert.assertEquals(-1, ledgerContext.compareTo(addressIsInvalid).toLong())
        Assert.assertEquals(1, ledgerContext.compareTo(addressIsExist).toLong())
    }

    @Test
    fun compareToWhenStartValueAddressIsExist() {
        val ledgerContext = LedgerContext(
            address = dummyAddress,
            ledgerProofStatus = LedgerProofStatus.ADDRESS_EXISTS
        )
        Assert.assertEquals(-1, ledgerContext.compareTo(addressDoesntExist).toLong())
        Assert.assertEquals(-1, ledgerContext.compareTo(addressIsInvalid).toLong())
    }

    @Test
    fun compareToWhenStartValueAddressIsExistAndSignatureIndexLess() {
        val ledgerContext = LedgerContext(
            address = dummyAddress,
            ledgerProofStatus = LedgerProofStatus.ADDRESS_EXISTS,
            ledgerValue = LedgerValue(0L, 0L, 1L)
        )
        val addressIsExist = LedgerContext(
            address = dummyAddress,
            ledgerProofStatus = LedgerProofStatus.ADDRESS_EXISTS,
            ledgerValue = LedgerValue(0L, 0L, 0L)
        )
        Assert.assertEquals(-1, ledgerContext.compareTo(addressIsExist).toLong())
    }

    @Test
    fun compareToWhenStartValueAddressIsExistAndSignatureIndexMore() {
        val ledgerContext = LedgerContext(
            address = dummyAddress,
            ledgerProofStatus = LedgerProofStatus.ADDRESS_EXISTS,
            ledgerValue = LedgerValue(0L, 0L, 1L)
        )
        val addressIsExist = LedgerContext(
            address = dummyAddress,
            ledgerProofStatus = LedgerProofStatus.ADDRESS_EXISTS,
            ledgerValue = LedgerValue(0L, 0L, 3L)
        )
        Assert.assertEquals(1, ledgerContext.compareTo(addressIsExist).toLong())
    }

    @Test
    fun compareToWhenAddressIsExistAndSignatureIndexEqualButUnitsLess() {
        val ledgerContext = LedgerContext(
            address = dummyAddress,
            ledgerProofStatus = LedgerProofStatus.ADDRESS_EXISTS,
            ledgerValue = LedgerValue(100L, 0L, 3L)
        )
        val addressIsExist = LedgerContext(
            address = dummyAddress,
            ledgerProofStatus = LedgerProofStatus.ADDRESS_EXISTS,
            ledgerValue = LedgerValue(0L, 0L, 3L)
        )
        Assert.assertEquals(-1, ledgerContext.compareTo(addressIsExist).toLong())
    }

    @Test
    fun compareToWhenAddressIsExistAndSignatureIndexEqualButUnitsMore() {
        val ledgerContext = LedgerContext(
            address = dummyAddress,
            ledgerProofStatus = LedgerProofStatus.ADDRESS_EXISTS,
            ledgerValue = LedgerValue(100L, 0L, 3L)
        )
        val addressIsExist = LedgerContext(
            address = dummyAddress,
            ledgerProofStatus = LedgerProofStatus.ADDRESS_EXISTS,
            ledgerValue = LedgerValue(101L, 0L, 3L)
        )
        Assert.assertEquals(1, ledgerContext.compareTo(addressIsExist).toLong())
    }

    @Test
    fun compareToWhenAddressIsExistAndSignatureIndexEqual() {
        val ledgerContext = LedgerContext(
            address = dummyAddress,
            ledgerProofStatus = LedgerProofStatus.ADDRESS_EXISTS,
            ledgerValue = LedgerValue(100L, 0L, 3L)
        )
        val addressIsExist = LedgerContext(
            address = dummyAddress,
            ledgerProofStatus = LedgerProofStatus.ADDRESS_EXISTS,
            ledgerValue = LedgerValue(100L, 0L, 3L)
        )
        Assert.assertEquals(-1, ledgerContext.compareTo(addressIsExist).toLong())
    }
}
